### DISCUSSION TITLE
How to write custom error message

### DISCUSSION THREAD

**Discussion by shujaatsiddiqui** [microsoft/RulesEngine#180] [Q&A]
I want to concatenate property value into error message. I have a list of collection and I want to concatenate what item in the list satisfy the expression criteria.

```
{
        "RuleName": "CheckforVTPumpSelectionErrors121-03",
        "ErrorMessage": "WARNING: 121 - VT PUMP Manual Pick Required for {0} HP {1} pump!",
         "Expression": "ProductFamily.PumpConfigurations.Any(r => r.PumpDescription == \"Water Feature\" || r.PumpDescription == \"Transfer\") && Flag1 == false"
}
```

I want to replace {0} with horse power value. 

**My DTO:**
```
public class ProductFamily
    {
        public int SCCR = 30;
        public List<PumpConfiguration> PumpConfigurations { get; set; }
    }

 public class PumpConfiguration
    {
        public PumpConfiguration()
        {
        }
        public string PumpDescription { get; internal set; }
        public int HP { get; internal set; }
    }
```

**Comment by abbasc52** [comment]
Hi @shujaatsiddiqui 
We partially support writing custom error message.  The syntax involves using `$()` around the variable. For e.g. `$(input1)`
you can find sample rule here:-
https://github.com/microsoft/RulesEngine/blob/bafbff281d16172bd7b4cf4c7785e684cfa2be2c/test/RulesEngine.UnitTest/TestData/rules4.json#L8

In your case you could use a local param to get the list item you want to print and then use it in error message.

Note: we only support basic parsing of error message, If you need more complex operations prefer using `OutputExpression` Action - https://microsoft.github.io/RulesEngine/#outputexpression

**Reply by shujaatsiddiqui** [reply]
![image](https://user-images.githubusercontent.com/51877809/124719844-01ad9500-df21-11eb-8764-0af474b6f0c2.png)

It is not replacing it with properties values. I ran your test case from samples.

**Comment by ElanHasson** [comment]
I was able to accomplish similar with a custom action and a dynamic lambda parser on the context values

**Reply by shujaatsiddiqui** [reply]
can you please share some examples? or sample code?
